In [21]:
import numpy as np
import time

In [22]:
X=[0.5,2.5]
Y=[0.2,0.9]


In [23]:
def sigmoid(w,b,x):
    return 1/(1+(np.exp(-(w*x-b))))

In [24]:
def gradW(w,b,x,y):
    fx = sigmoid(w,b,x)
    return (fx - y) * (fx) * (1 - fx) * x

In [25]:
def gradB(w,b,x,y):
    fx = sigmoid(w,b,x)
    return (fx - y) * (fx) * (1 - fx)

In [26]:
def lossFuction(w,b):
    error = 0
    for (dx,dy) in zip(X,Y):
        fx = sigmoid(w,b,dx)
        error += 0.5 * (fx - dy)**2
    return error

In [31]:
def vanila_grad(w,b,learning_rate,epochs):
    seconds = time.time()
    for i in range(epochs):
        dw,db = 0,0
        for (x,y) in zip(X,Y):
            dw += gradW(w,b,x,y)
            db += gradB(w,b,x,y)
        w = w - learning_rate * dw
        b = b - learning_rate * db
        print(lossFuction(w,b))
    print("Seconds since epoch =", time.time() - seconds)

In [32]:
def moumentum_grad(w,b,learning_rate,epochs):
    seconds = time.time()
    gama = 0.1
    w_pre,b_pre = 0,0
    for i in range(epochs):
        dw,db = 0,0
        for (x,y) in zip(X,Y):
            dw += gradW(w,b,x,y)
            db += gradB(w,b,x,y)
        w = w - ((gama * w_pre) +(learning_rate * dw))
        b = b - ((gama * b_pre) +(learning_rate * db))
        w_pre = w
        b_pre = b
        print(lossFuction(w,b))
    print("Seconds since epoch =", time.time() - seconds)

In [33]:
vanila_grad(-2,-2,0.1,100)
print("-----------------------------")
moumentum_grad(-2,-2,0.1,100)


0.5046873188453432
0.5048992456379796
0.5050876025573474
0.5052512236473746
0.505388878376643
0.5054992670295545
0.5055810157331817
0.5056326710907652
0.5056526943908577
0.5056394553591271
0.5055912254178838
0.5055061704165345
0.505382342794479
0.5052176731365383
0.5050099610799818
0.5047568655317709
0.5044558941549739
0.5041043920846991
0.5036995298367204
0.5032382903766361
0.5027174553244939
0.5021335902800017
0.5014830292676171
0.5007618583199975
0.49996589824384485
0.4990906866456812
0.4981314593384962
0.49708313130589565
0.4959402774711194
0.4946971136074841
0.4933474778382625
0.4918848133122474
0.4903021528112296
0.48859210625279914
0.4867468523019164
0.48475813560307657
0.4826172714963104
0.480315160487618
0.4778423152074459
0.47518890310389655
0.4723448086669465
0.46929971954071203
0.46604324141121367
0.4625650469938723
0.45885506469731924
0.45490371248411665
0.45070218192404216
0.44624277624484754
0.441519304104135
0.4365275276018815
0.4312656585238414
0.4257348908378943
0.419

<table>
    <tr>
        <th>Weight</th>
        <th>Bais</th>
        <th>Learning rate</th>
        <th>epoch</th>
        <th>Vanila-Gradient-Time</th>
        <th>Momentum-Gradient-Time</th>
        <th>Vanila-Gradient-Loss</th>
        <th>Momentum-Gradient-Loss</th>
    </tr>
    <tr>
        <td>-2</td>
        <td>-2</td>
        <td>0.1</td>
        <td>10</td>
        <td>0.0005846023559570312</td>
        <td>0.00041937828063964844</td>
        <td>0.5056394553591271</td>
        <td>0.26750690145032485</td>
    </tr>
      <tr>
        <td>-2</td>
        <td>-2</td>
        <td>0.1</td>
        <td>50</td>
        <td>0.002499103546142578</td>
        <td>0.0017952919006347656</td>
        <td>0.4365275276018815</td>
        <td>0.09947335503507101</td>
    </tr> 
    <tr>
        <td>-2</td>
        <td>-2</td>
        <td>0.1</td>
        <td>75</td>
        <td>0.0009799003601074219</td>
        <td>0.0008487701416015625</td>
        <td>0.28238180789995837</td>
        <td>0.09864978585459042</td>
    </tr>
    <tr>
        <td>-2</td>
        <td>-2</td>
        <td>0.1</td>
        <td>100</td>
        <td>0.00393986701965332</td>
        <td>0.010802745819091797</td>
        <td>0.2428709093366296</td>
        <td>0.09859690650726119</td>
    </tr>
</table>
